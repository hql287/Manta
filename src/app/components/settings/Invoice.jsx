// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';
const ipc = require('electron').ipcRenderer;

// Custom Libs
const openDialog = require('../../renderers/dialog.js');
import _withFadeInAnimation from '../shared/hoc/_withFadeInAnimation';

import Currency from './_partials/invoice/Currency';
import Fields from './_partials/invoice/Fields';
import Other from './_partials/invoice/Other';
import Tax from './_partials/invoice/Tax';

// Component
class Invoice extends Component {
  constructor(props) {
    super(props);
    this.state = this.props.invoice;
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleTaxChange = this.handleTaxChange.bind(this);
    this.handleCurrencyChange = this.handleCurrencyChange.bind(this);
    this.handleVisibilityChange = this.handleVisibilityChange.bind(this);
  }

  componentDidMount() {
    const { t } = this.props;
    ipc.on('no-access-directory', (event, message) => {
      openDialog({
        type: 'warning',
        title: t('dialog:noAccess:title'),
        message: `${message}. ${t('dialog:noAccess:message')}`,
      });
    });

    ipc.on('confirmed-export-directory', (event, path) => {
      this.setState({ exportDir: path }, () => {
        this.props.updateSettings('invoice', this.state);
      });
    });
  }

  componentWillUnmount() {
    ipc.removeAllListeners('no-access-directory');
    ipc.removeAllListeners('confirmed-export-directory');
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    this.setState({ [name]: value }, () => {
      this.props.updateSettings('invoice', this.state);
    });
  }

  handleTaxChange(event) {
    const target = event.target;
    const name = target.name;
    const value = name === 'amount' ? parseFloat(target.value) : target.value;
    this.setState(
      {
        tax: Object.assign({}, this.state.tax, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  handleCurrencyChange(event) {
    const target = event.target;
    const name = target.name;
    const value = name === 'fraction' ? parseInt(target.value, 10) : target.value;
    this.setState(
      {
        currency: Object.assign({}, this.state.currency, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  handleVisibilityChange(event) {
    const target = event.target;
    const name = target.name;
    const value = target.checked;
    this.setState(
      {
        required_fields: Object.assign({}, this.state.required_fields, {
          [name]: value,
        }),
      },
      () => {
        this.props.updateSettings('invoice', this.state);
      }
    );
  }

  selectExportDir() {
    ipc.send('select-export-directory');
  }

  render() {
    const { t } = this.props;
    const {
      exportDir,
      template,
      currency,
      tax,
      required_fields,
      dateFormat,
    } = this.state;
    return [
      <Fields
        key="required_fields_settings"
        required_fields={required_fields}
        handleVisibilityChange={this.handleVisibilityChange}
        t={t}
      />,
      <Tax
        key="tax_settings"
        handleTaxChange={this.handleTaxChange}
        tax={tax}
        t={t}
      />,
      <Currency
        key="currency_settings"
        currency={currency}
        handleCurrencyChange={this.handleCurrencyChange}
        t={t}
      />,
      <Other
        key="other_settings"
        dateFormat={dateFormat}
        exportDir={exportDir}
        template={template}
        handleInputChange={this.handleInputChange}
        selectExportDir={this.selectExportDir}
        t={t}
      />
    ];
  }
}

Invoice.propTypes = {
  invoice: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired,
  updateSettings: PropTypes.func.isRequired,
};

export default _withFadeInAnimation(Invoice);
