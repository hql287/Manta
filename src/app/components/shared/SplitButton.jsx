// Libraries
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Styles
import styled from 'styled-components';

const Wrapper = styled.div`
  position: relative;
`;

const ButtonsGroup = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  > a {
    flex: 1;
    border: 1px solid rgba(0, 0, 0, 0.1);
    display: flex;
    justify-content: center;
    // align-items: center;
    padding: 4px 0;
    font-size: 12px;
    text-decoration: none;
    background: #ffffff;
    text-transform: uppercase;
    letter-spacing: 1px;
    color: #000;
    &:first-child {
      border-radius: 4px 0 0 4px;
      padding: 4px 12px;
      text-align: left;
      justify-content: flext-start;
      // align-items: flext-start;
    }
    &:last-child {
      border-radius: 0 4px 4px 0;
      border-left: none;
      max-width: 24px;
    }
  }
`;

const Addon = styled.div`
  &:before {
    content: ' ';
    height: 10px;
    width: 10px;
    position: absolute;
    top: -6px;
    right: 10px;
    transform: rotate(45deg);
    background: white;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-left: 1px solid rgba(0, 0, 0, 0.1);
  }
  display: flex;
  flex-direction: column;
  position: absolute;
  right: 0;
  top: 42px;
  width: 100%;
  background: white;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.1);
  z-index: +999;
  a {
    text-transform: uppercase;
    font-size: 12px;
    letter-spacing: 1px;
    padding: 6px 12px;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    color: black;
    width: 100%;
    text-align: left;
    z-index: 10;
    &:hover {
      text-decoration: none;
      background: #f9fafa;
    }
    &:last-child {
      border-bottom: none;
    }
  }
`;

class SplitButton extends Component {
  constructor(props) {
    super(props);
    this.state = { showOptions: false };
    this.handleMainAction = this.handleMainAction.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.renderListItems = this.renderListItems.bind(this);
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.state.showOptions !== nextState.showOptions) {
      return true;
    }
    return false;
  }

  handleClick() {
    if (!this.state.showOptions) {
      document.addEventListener('click', this.handleOutsideClick, false);
    } else {
      document.removeEventListener('click', this.handleOutsideClick, false);
    }
    this.setState({ showOptions: !this.state.showOptions });
  }

  handleOutsideClick(e) {
    if (this.node.contains(e.target)) {
      return;
    }
    this.handleClick();
  }

  handleMainAction(e) {
    this.setState({ showOptions: false }, () => {
      this.props.mainButton.action();
    });
  }

  renderListItems() {
    const { options } = this.props;
    return options.map(option => (
      <a
        key={option.label.toString()}
        href="#"
        onClick={() => {
          option.action();
          this.handleClick();
        }}
      >
        {option.label}
      </a>
    ));
  }

  render() {
    return (
      <div ref={node => (this.node = node)}>
        <Wrapper>
          <ButtonsGroup>
            <a href="#" onClick={this.handleMainAction}>
              {this.props.mainButton.label}
            </a>
            <a href="#" onClick={this.handleClick}>
              <i className="ion-arrow-down-b" />
            </a>
          </ButtonsGroup>
          {this.state.showOptions && <Addon>{this.renderListItems()}</Addon>}
        </Wrapper>
      </div>
    );
  }
}

SplitButton.propTypes = {
  mainButton: PropTypes.shape({
    label: PropTypes.string.isRequired,
    action: PropTypes.func.isRequired,
  }).isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      action: PropTypes.func.isRequired,
    })
  ).isRequired,
};

export default SplitButton;
