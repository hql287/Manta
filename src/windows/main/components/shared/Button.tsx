// Libraries
import React, { FC, ReactNode } from 'react'
import styled from 'styled-components'

interface ButtonProps {
  link?: boolean
  danger?: boolean
  onClick?: () => void
  primary?: boolean
  success?: boolean
  children: ReactNode
  block?: boolean
  active?: boolean
}

const ButtonStyle = styled.button<ButtonProps>`
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  padding: 4px 15px;
  font-size: 12px;
  text-decoration: none;
  background: #ffffff;
  border: 1px solid #e0e1e1;
  text-transform: uppercase;
  letter-spacing: 1px;
  ${(props) => props.block && `width: 100%;`}
  ${(props) =>
    props.primary &&
    `
    background: #469fe5;
    color: white;
  `}
  ${(props) =>
    props.success &&
    `
    background: #6bbb69;
    color: white;
  `}
  ${(props) =>
    props.danger &&
    `
    background: #EC476E;
    color: white;
  `}
  ${(props) =>
    props.active &&
    `
    background: #F2F3F4;
    color: #4F555C;
  `}
  &:hover {
    cursor: pointer;
    text-decoration: none;
  }
`

const ButtonLinkStyle = styled.button<ButtonProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  background: none;
  border: none;
  padding: 0;
  margin: 0;
  ${(props) => props.primary && `color: #469fe5;`}
  ${(props) => props.success && `color: #6bbb69;`}
  ${(props) => props.danger && `color: #EC476E;`}
  &:hover {
    cursor: pointer;
  }
`

const ButtonsGroupStyle = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  > button {
    margin: 0 !important;
    border-radius: 0;
    &:first-child {
      border-radius: 4px 0 0 4px;
    }
    &:last-child {
      border-radius: 0 4px 4px 0;
    }
    &:not(:first-child) {
      border-left: 0;
    }
  }
`

export const Button: FC<ButtonProps> = (props) => {
  return props.link ? (
    <ButtonLinkStyle {...props}>{props.children}</ButtonLinkStyle>
  ) : (
    <ButtonStyle {...props}>{props.children}</ButtonStyle>
  )
}

export const ButtonsGroup: FC<{ children: ReactNode }> = ({ children }) => (
  <ButtonsGroupStyle>{children}</ButtonsGroupStyle>
)
