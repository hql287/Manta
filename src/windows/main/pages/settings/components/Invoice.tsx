// Libraries
import React, { useEffect, useState } from 'react'
//import { ipcRenderer as ipc } from 'electron'

// Custom Libs
//import openDialog from '../../renderers/dialog.js'
import _withFadeInAnimation from '@uiSharedComponents/hoc/_withFadeInAnimation'

import Currency from './_partials/invoice/Currency'
import Fields from './_partials/invoice/Fields'
import Other from './_partials/invoice/Other'
import Tax from './_partials/invoice/Tax'

const { ipcRenderer: ipc, createDialogWindow } = window.electronAPI

interface TaxState {
  name: string
  amount: number
}

interface CurrencyState {
  code: string
  separator: string
  placement: string
  fraction: number
}

interface RequiredFields {
  [key: string]: boolean
}

interface InvoiceState {
  exportDir: string
  template: string
  currency: CurrencyState
  tax: TaxState
  required_fields: RequiredFields
  dateFormat: string
}

interface InvoiceProps {
  invoice: InvoiceState
  t: (key: string) => string
  updateSettings: (field: string, data: InvoiceState) => void
}

const Invoice: React.FC<InvoiceProps> = ({ invoice, t, updateSettings }) => {
  const [state, setState] = useState<InvoiceState>(invoice)

  const noAccessHandler = (
    _event: Electron.IpcRendererEvent,
    message: string
  ) => {
    createDialogWindow({
      type: 'warning',
      title: t('dialog:noAccess:title'),
      message: `${message}. ${t('dialog:noAccess:message')}`,
    })
  }

  const confirmedExportHandler = (
    _event: Electron.IpcRendererEvent,
    path: string
  ) => {
    setState((prevState) => {
      const updatedState = { ...prevState, exportDir: path }
      updateSettings('invoice', updatedState)
      return updatedState
    })
  }

  useEffect(() => {
    ipc.on('no-access-directory', noAccessHandler)
    ipc.on('confirmed-export-directory', confirmedExportHandler)

    return () => {
      ipc.removeListener('no-access-directory', noAccessHandler)
      ipc.removeListener('confirmed-export-directory', confirmedExportHandler)
    }
  }, [t, updateSettings])

  const handleInputChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { name, value, type, checked } = event.target
    const updatedValue = type === 'checkbox' ? checked : value

    setState((prevState) => {
      const updatedState = { ...prevState, [name]: updatedValue }
      updateSettings('invoice', updatedState)
      return updatedState
    })
  }

  const handleTaxChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { name, value } = event.target
    const updatedValue = name === 'amount' ? parseFloat(value) : value

    setState((prevState) => {
      const updatedTax = { ...prevState.tax, [name]: updatedValue }
      const updatedState = { ...prevState, tax: updatedTax }
      updateSettings('invoice', updatedState)
      return updatedState
    })
  }

  const handleCurrencyChange = (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ) => {
    const { name, value } = event.target
    const updatedValue = name === 'fraction' ? parseInt(value, 10) : value

    setState((prevState) => {
      const updatedCurrency = { ...prevState.currency, [name]: updatedValue }
      const updatedState = { ...prevState, currency: updatedCurrency }
      updateSettings('invoice', updatedState)
      return updatedState
    })
  }

  const handleVisibilityChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const { name, checked } = event.target

    setState((prevState) => {
      const updatedFields = { ...prevState.required_fields, [name]: checked }
      const updatedState = { ...prevState, required_fields: updatedFields }
      updateSettings('invoice', updatedState)
      return updatedState
    })
  }

  const selectExportDir = () => {
    ipc.send('select-export-directory')
  }

  const { exportDir, template, currency, tax, required_fields, dateFormat } =
    state

  return (
    <>
      <Fields
        required_fields={required_fields}
        handleVisibilityChange={handleVisibilityChange}
        t={t}
      />
      <Tax handleTaxChange={handleTaxChange} tax={tax} t={t} />
      <Currency
        currency={currency}
        handleCurrencyChange={handleCurrencyChange}
        t={t}
      />
      <Other
        dateFormat={dateFormat}
        exportDir={exportDir}
        template={template}
        handleInputChange={handleInputChange}
        selectExportDir={selectExportDir}
        t={t}
      />
    </>
  )
}

export default _withFadeInAnimation(Invoice)
