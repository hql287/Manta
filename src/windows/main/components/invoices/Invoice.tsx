// Libraries
import React from 'react'
import { truncate } from 'lodash'
import styled from 'styled-components'
import moment from 'moment'
//import { ipcRenderer as ipc } from 'electron'

const { ipcRenderer: ipc } = window.electronAPI

// Helper
import { formatNumber } from '@uiHelpers/formatNumber'
import { calTermDate } from '@uiHelpers/date'

// Custom Components
import { Button } from '@uiSharedComponents/Button'
import SplitButton from '@uiSharedComponents/SplitButton'

// Styled Components
const Wrapper = styled.div`
  position: relative;
  border: 1px solid rgba(0, 0, 0, 0.1);
  background: white;
  margin-bottom: 30px;
  border-radius: 4px;
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
`

const Header = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 25px 30px;
`

const StatusBar = styled.div<{ status: string }>`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 6px;
  border-radius: 4px 4px 0 0;
  ${(props) => props.status === 'pending' && `background: #469FE5;`}
  ${(props) => props.status === 'paid' && `background: #6BBB69;`}
  ${(props) => props.status === 'refunded' && `background: #4F555C;`}
  ${(props) => props.status === 'cancelled' && `background: #EC476E;`}
`

const Status = styled.div<{ status: string }>`
  font-size: 14px;
  font-weight: 600;
  text-transform: uppercase;
  letter-spacing: 1px;
  ${(props) => props.status === 'pending' && `color: #469FE5;`}
  ${(props) => props.status === 'paid' && `color: #6BBB69;`}
  ${(props) => props.status === 'refunded' && `color: #4F555C;`}
  ${(props) => props.status === 'cancelled' && `color: #EC476E;`}
  span {
    display: flex;
    align-items: center;
    i {
      margin-right: 5px;
    }
  }
  i.ion-checkmark,
  i.ion-loop,
  i.ion-backspace,
  i.ion-arrow-return-left {
    font-size: 18px;
    line-height: 18px;
  }
`

const ButtonsGroup = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-end;
  i {
    margin-left: 10px;
    color: #b4b7ba;
  }
  i.ion-trash-a {
    font-size: 24px;
    line-height: 24px;
    &:hover {
      color: #ec476e;
    }
  }
  i.ion-ios-copy {
    font-size: 24px;
    line-height: 24px;
    &:hover {
      color: #469fe5;
    }
  }
`

const Body = styled.div`
  padding: 0 30px;
`

const Footer = styled.div`
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  padding: 20px;
  display: flex;
  justify-content: space-between;
  align-items: center;
  > * {
    flex: 1;
    margin: 0 10px;
    &:first-child {
      flex: 3;
    }
  }
`

const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-around;
`

const Field = styled.div`
  position: relative;
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  margin-bottom: 15px;
  text-transform: capitalize;
  h2 {
    font-size: 21px;
    color: #283641;
    margin-bottom: 0;
    font-weight: 500;
  }
  label {
    font-size: 11px;
    color: #b4b7ba;
    text-transform: uppercase;
    letter-spacing: 1px;
    margin-bottom: 4px;
    font-weight: 400;
  }
  p {
    font-weight: 300;
    margin-bottom: 0px;
    font-size: 14px;
  }
`

// Types
interface Currency {
  placement: 'before' | 'after'
  code: string
  fraction: number
  separator: string
}

interface DueDate {
  useCustom: boolean
  paymentTerm: string
  selectedDate?: string
}

interface Recipient {
  fullname: string
}

interface InvoiceProps {
  invoice: {
    _id: string
    recipient: Recipient
    currency: Currency
    tax?: number
    grandTotal: number
    status: 'pending' | 'paid' | 'refunded' | 'cancelled'
    dueDate: DueDate
    created_at: string
    invoiceID?: string
    configs?: {
      dateFormat: string
    }
  }
  dateFormat: string
  deleteInvoice: (id: string) => void
  duplicateInvoice: (invoice: InvoiceProps['invoice']) => void
  editInvoice: (invoice: InvoiceProps['invoice']) => void
  setInvoiceStatus: (id: string, status: string) => void
  t: (key: string) => string
}

const Invoice: React.FC<InvoiceProps> = ({
  invoice,
  dateFormat,
  deleteInvoice,
  duplicateInvoice,
  editInvoice,
  setInvoiceStatus,
  t,
}) => {
  const viewInvoice = () => {
    ipc.send('preview-invoice', invoice)
  }

  const renderStatus = () => {
    switch (invoice.status) {
      case 'cancelled':
        return (
          <span>
            <i className="ion-backspace" />
            {t('invoices:status:cancelled')}
          </span>
        )
      case 'paid':
        return (
          <span>
            <i className="ion-checkmark" />
            {t('invoices:status:paid')}
          </span>
        )
      case 'refunded':
        return (
          <span>
            <i className="ion-arrow-return-left" />
            {t('invoices:status:refunded')}
          </span>
        )
      default:
        return (
          <span>
            <i className="ion-loop" />
            {t('invoices:status:pending')}
          </span>
        )
    }
  }

  const renderDueDate = () => {
    const { dueDate } = invoice
    const format = invoice.configs?.dateFormat || dateFormat

    if (dueDate.useCustom) {
      return moment(dueDate.selectedDate).format(format)
    }

    const paymentTermDate = calTermDate(invoice.created_at, dueDate.paymentTerm)
    return `
      ${t(`form:fields:dueDate:paymentTerms:${dueDate.paymentTerm}:label`)}
      (
      ${moment(paymentTermDate).format(format)}
      )
    `
  }

  const statusActions = [
    {
      label: t('invoices:status:pending'),
      action: () => setInvoiceStatus(invoice._id, 'pending'),
    },
    {
      label: t('invoices:status:refunded'),
      action: () => setInvoiceStatus(invoice._id, 'refunded'),
    },
    {
      label: t('invoices:status:cancelled'),
      action: () => setInvoiceStatus(invoice._id, 'cancelled'),
    },
  ]

  const currencyBefore = invoice.currency.placement === 'before'

  return (
    <div className="col-lg-6">
      <Wrapper>
        <StatusBar status={invoice.status} />
        <Header>
          <Status status={invoice.status}>{renderStatus()}</Status>
          <ButtonsGroup>
            <Button link onClick={() => duplicateInvoice(invoice)}>
              <i className="ion-ios-copy" />
            </Button>
            <Button link onClick={() => deleteInvoice(invoice._id)}>
              <i className="ion-trash-a" />
            </Button>
          </ButtonsGroup>
        </Header>
        <Body>
          <Row>
            <Field>
              <label>{t('invoices:fields:client')}</label>
              <h2>{invoice.recipient.fullname}</h2>
            </Field>
          </Row>
          <Row>
            <Field>
              <label>{t('invoices:fields:invoiceID')}</label>
              <p>
                {invoice.invoiceID
                  ? invoice.invoiceID
                  : truncate(invoice._id, { length: 8, omission: '' })}
              </p>
            </Field>
            <Field>
              <label>{t('invoices:fields:total')}</label>
              <p>
                {currencyBefore ? invoice.currency.code : null}{' '}
                {formatNumber(
                  invoice.grandTotal,
                  invoice.currency.fraction,
                  invoice.currency.separator
                )}{' '}
                {currencyBefore ? null : invoice.currency.code}
              </p>
            </Field>
          </Row>
          <Row>
            <Field>
              <label>{t('invoices:fields:createdDate')}</label>
              <p>{moment(invoice.created_at).format(dateFormat)}</p>
            </Field>
            <Field>
              <label>{t('invoices:fields:dueDate')}</label>
              <p>{invoice.dueDate && renderDueDate()}</p>
            </Field>
          </Row>
        </Body>
        <Footer>
          <SplitButton
            mainButton={{
              label: t('invoices:btns:markAsPaid'),
              action: () => setInvoiceStatus(invoice._id, 'paid'),
            }}
            options={statusActions}
          />
          <Button onClick={() => editInvoice(invoice)}>
            {t('invoices:btns:edit')}
          </Button>
          <Button onClick={viewInvoice}>{t('invoices:btns:view')}</Button>
        </Footer>
      </Wrapper>
    </div>
  )
}

export default Invoice
