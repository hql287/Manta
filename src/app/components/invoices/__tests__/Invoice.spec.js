// Libs
import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// Component
import Invoice from '../Invoice';
import Button from '../../shared/Button';
import SplitButton from '../../shared/SplitButton';

// Mocks
const invoice = {
  _id: '37b2804e-bfc1-4289-b1d7-226c5652ac91',
  status: 'pending',
  currency: {
    code: 'USD',
    decimal_digits: 2,
    name: 'US Dollar',
    name_plural: 'US dollars',
    rounding: 0,
    symbol: '$',
    symbol_native: '$',
  },
  subtotal: 3843,
  grandTotal: 3843,
  recipient: {
    _id: '3bd85cb9-7675-4d59-b9c3-305481cb77c9',
    company: 'Lindgren Group',
    email: 'Hattie42@gmail.com',
    fullname: 'Penelope Hettinger',
    phone: '854-306-7837',
  },
};

const editInvoice = jest.fn();
const deleteInvoice = jest.fn();
const duplicateInvoice = jest.fn();
const setInvoiceStatus = jest.fn();
const dateFormat = 'MM/DD/YY';
const currencyPlacement = 'before';
Date.now = jest.genMockFunction().mockReturnValue(0);

const t = jest.fn(status => {
  switch (status) {
    case 'invoices:status:cancelled': {
      return 'Cancelled';
    }
    case 'invoices:status:refunded': {
      return 'Refunded';
    }
    case 'invoices:status:paid': {
      return 'Paid';
    }
    default: {
      return 'Pending';
    }
  }
});

// Tests
describe('Renders correctly to the DOM', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Invoice
        t={t}
        invoice={invoice}
        editInvoice={editInvoice}
        deleteInvoice={deleteInvoice}
        duplicateInvoice={duplicateInvoice}
        setInvoiceStatus={setInvoiceStatus}
        dateFormat={dateFormat}
        currencyPlacement={currencyPlacement}
      />
    );
  });

  it('receives correct props', () => {
    const mountWrapper = mount(
      <Invoice
        t={t}
        invoice={invoice}
        editInvoice={editInvoice}
        deleteInvoice={deleteInvoice}
        duplicateInvoice={duplicateInvoice}
        setInvoiceStatus={setInvoiceStatus}
        dateFormat={dateFormat}
        currencyPlacement={currencyPlacement}
      />
    );
    expect(mountWrapper.prop('invoice')).toEqual(invoice);
    expect(mountWrapper.prop('deleteInvoice')).toEqual(deleteInvoice);
    expect(mountWrapper.prop('setInvoiceStatus')).toEqual(setInvoiceStatus);
  });

  it('render with a class of .col-lg-6', () => {
    expect(wrapper.find('.col-lg-6')).toHaveLength(1);
    expect(wrapper.find('.col-lg-6')).not.toHaveLength(2);
  });

  it('render invoice status in the header', () => {
    expect(
      wrapper.find('Invoice__Header').find('Invoice__Status')
    ).toHaveLength(1);
  });

  it('render invoice status correctly', () => {
    // Pending Invoice (default)
    expect(
      wrapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text()
    ).toEqual('Pending');

    // Paid Invoice
    const paidInvoice = Object.assign({}, invoice, { status: 'paid' });
    const paidInvoiceWapper = shallow(
      <Invoice
        t={t}
        language="en"
        invoice={paidInvoice}
        editInvoice={editInvoice}
        deleteInvoice={deleteInvoice}
        duplicateInvoice={duplicateInvoice}
        setInvoiceStatus={setInvoiceStatus}
        dateFormat={dateFormat}
        currencyPlacement={currencyPlacement}
      />
    );
    expect(
      paidInvoiceWapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text()
    ).toEqual('Paid');

    // Cancelled Invoice
    const cancelledInvoice = Object.assign({}, invoice, {
      status: 'cancelled',
    });
    const cancelledInvoiceWapper = shallow(
      <Invoice
        t={t}
        language="en"
        invoice={cancelledInvoice}
        editInvoice={editInvoice}
        deleteInvoice={deleteInvoice}
        duplicateInvoice={duplicateInvoice}
        setInvoiceStatus={setInvoiceStatus}
        dateFormat={dateFormat}
        currencyPlacement={currencyPlacement}
      />
    );
    expect(
      cancelledInvoiceWapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text()
    ).toEqual('Cancelled');

    // Refunded Invoice
    const refundedInvoice = Object.assign({}, invoice, { status: 'refunded' });
    const refundedInvoiceWapper = shallow(
      <Invoice
        t={t}
        language="en"
        invoice={refundedInvoice}
        editInvoice={editInvoice}
        deleteInvoice={deleteInvoice}
        duplicateInvoice={duplicateInvoice}
        setInvoiceStatus={setInvoiceStatus}
        dateFormat={dateFormat}
        currencyPlacement={currencyPlacement}
      />
    );
    expect(
      refundedInvoiceWapper
        .find('Invoice__Header')
        .find('Invoice__Status')
        .find('span')
        .text()
    ).toEqual('Refunded');
  });

  it('render delete and duplicate button in the header', () => {
    expect(wrapper.find('Invoice__Header').find(Button)).toHaveLength(2);
  });

  it('render a split button and 2 buttons in the footer', () => {
    expect(wrapper.find('Invoice__Footer').find(SplitButton)).toHaveLength(1);
    expect(wrapper.find('Invoice__Footer').find(Button)).toHaveLength(2);
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <Invoice
          t={t}
          invoice={invoice}
          editInvoice={editInvoice}
          deleteInvoice={deleteInvoice}
          duplicateInvoice={duplicateInvoice}
          setInvoiceStatus={setInvoiceStatus}
          dateFormat={dateFormat}
          currencyPlacement={currencyPlacement}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('handle clicks correctly', () => {
  let wrapper,
    editBtn,
    viewBtn,
    duplicateBtn,
    deleteBtn,
    spyViewAction,
    spyEditAction,
    spyDuplicateAction;
  beforeEach(() => {
    spyViewAction = jest.spyOn(Invoice.prototype, 'viewInvoice');
    spyEditAction = jest.spyOn(Invoice.prototype, 'editInvoice');
    spyDuplicateAction = jest.spyOn(Invoice.prototype, 'duplicateInvoice');
    wrapper = shallow(
      <Invoice
        t={t}
        invoice={invoice}
        editInvoice={editInvoice}
        deleteInvoice={deleteInvoice}
        duplicateInvoice={duplicateInvoice}
        setInvoiceStatus={setInvoiceStatus}
        dateFormat={dateFormat}
        currencyPlacement={currencyPlacement}
      />
    );
    viewBtn = wrapper
      .find('Invoice__Footer')
      .find(Button)
      .last();
    editBtn = wrapper
      .find('Invoice__Footer')
      .find(Button)
      .first();
    deleteBtn = wrapper.find(Button).at(1);
    duplicateBtn = wrapper.find(Button).first();
  });

  it('handle edit action correctly', () => {
    editBtn.simulate('click');
    expect(editInvoice).toHaveBeenCalled();
    expect(editInvoice).toHaveBeenCalledWith(invoice);
  });

  it('handle delete action correctly', () => {
    deleteBtn.simulate('click');
    expect(deleteInvoice).toHaveBeenCalled();
    expect(deleteInvoice).toHaveBeenCalledWith(invoice._id);
  });

  it('handle duplicate action correctly', () => {
    duplicateBtn.simulate('click');
    expect(duplicateInvoice).toHaveBeenCalled();
    expect(duplicateInvoice).toHaveBeenCalledWith(invoice);
  });

  it('handle view action correctly', () => {
    viewBtn.simulate('click');
    expect(spyViewAction).toHaveBeenCalled();
  });

  it('handle edit action correctly', () => {
    editBtn.simulate('click');
    expect(spyEditAction).toHaveBeenCalled();
  });
});
