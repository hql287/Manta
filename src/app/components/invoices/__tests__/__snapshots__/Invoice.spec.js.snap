// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly to the DOM matches snapshot 1`] = `
<div
  className="col-lg-6"
>
  <div
    className="Invoice__Wrapper-cjNJPS coWKLW"
  >
    <div
      className="Invoice__StatusBar-jBWtSa joKiG"
    />
    <div
      className="Invoice__Header-cGSpHV gmDFcR"
    >
      <div
        className="Invoice__Status-fxIELu bbALRd"
      >
        <span>
          <i
            className="ion-loop"
          />
          Pending
        </span>
      </div>
      <div
        className="Invoice__ButtonsGroup-PwNbx emusdh"
      >
        <button
          className="Button__ButtonLinkStyle-caVwCA imXwRc"
          onClick={[Function]}
        >
          <i
            className="ion-ios-copy"
          />
        </button>
        <button
          className="Button__ButtonLinkStyle-caVwCA imXwRc"
          onClick={[Function]}
        >
          <i
            className="ion-trash-a"
          />
        </button>
      </div>
    </div>
    <div
      className="Invoice__Body-eBItTh igYqzw"
    >
      <div
        className="Invoice__Row-hFNBjq cQdnct"
      >
        <div
          className="Invoice__Field-cZpgXH fDrpWO"
        >
          <label>
            Pending
          </label>
          <h2>
            Penelope Hettinger
          </h2>
        </div>
      </div>
      <div
        className="Invoice__Row-hFNBjq cQdnct"
      >
        <div
          className="Invoice__Field-cZpgXH fDrpWO"
        >
          <label>
            Pending
          </label>
          <p>
            37b2804e
          </p>
        </div>
        <div
          className="Invoice__Field-cZpgXH fDrpWO"
        >
          <label>
            Pending
          </label>
          <p>
             
            3 843
             
            USD
          </p>
        </div>
      </div>
      <div
        className="Invoice__Row-hFNBjq cQdnct"
      >
        <div
          className="Invoice__Field-cZpgXH fDrpWO"
        >
          <label>
            Pending
          </label>
          <p>
            01/01/70
          </p>
        </div>
        <div
          className="Invoice__Field-cZpgXH fDrpWO"
        >
          <label>
            Pending
          </label>
          <p />
        </div>
      </div>
    </div>
    <div
      className="Invoice__Footer-dcHDEd cJiGTY"
    >
      <div>
        <div
          className="SplitButton__Wrapper-jsRMIo gVGdJM"
        >
          <div
            className="SplitButton__ButtonsGroup-hdzsBq dhTQNh"
          >
            <a
              href="#"
              onClick={[Function]}
            >
              Pending
            </a>
            <a
              href="#"
              onClick={[Function]}
            >
              <i
                className="ion-arrow-down-b"
              />
            </a>
          </div>
        </div>
      </div>
      <button
        className="Button__ButtonStyle-coFSWz coAhhp"
        onClick={[Function]}
      >
        Pending
      </button>
      <button
        className="Button__ButtonStyle-coFSWz coAhhp"
        onClick={[Function]}
      >
        Pending
      </button>
    </div>
  </div>
</div>
`;
